### PR TITLE
LDAP authentication, user accounts

### PR DESCRIPTION
**Pull request by giuliojiang**
Before approving, please test if LDAP authentication works for you (using your college username & password)

You need IMPERIAL VPN to be able to connect to authentication server.

Compile using `make`

Start server using `./RunServer start --https`
use `albano` as temporary keystore password

Use Firefox to navigate to `https://localhost:4444` and accept the self-signed certificate.
Then navigate to `https://localhost:4443`

### CODE REVIEW

**Comment by AbraaoMota** [comment]
Tried to follow steps.
When running `./RunServer start --https` for the first time, it asked me some location based questions ( i think it was to generate the certificate) on the terminal. I went on `https://localhost:4444` and it didn't respond with any page at all. Going onto `https://localhost:4443`, it said the page was unsafe (on chrome) and the lock was crossed out. Rendered as normal. At no point was I greeted with the LDAP login (the one that CATe uses). What would be the steps to wipe the certificates to try again?

**Comment by giuliojiang** [comment]
this is known to be broken on Chrome, which doesn't allow you to accept a WebSocket signature at all. Using Firefox, if you navigate to :4444 it will ask you to accept the certificate, and then fail to load the page because that's just the websocket address.
If :4444 doesn't show anything at all, :4443 will not be able to connect to WebSocket.

There is no LDAP login popup, it uses the same login section as before.

**Comment by AbraaoMota** [comment]
Can't test on firefox for some reason, not loading over X11. ¯ \  _(ツ)_/¯
Tried on opera, navigating to :4444 did nothing, navigating to :4443 showed me a warning to accept a certificate, once I accepted it rendered the page, logging in was ineffective with real or fake details.

Does this mean we can only run ALABNO with LDAP on firefox?

**Comment by giuliojiang** [comment]
it means that as long as we use a self-signed certificate, only Firefox
When we use a verified certificate, any browser will work fine

**Comment by giuliojiang** [comment]
maybe I can deploy it somewhere useless and let you try the log-in. Or you can do it in a VirtualBox on your pc

**Comment by helicopter88** [comment]
Python reports that flask cannot be found, yet I pip installed it multiple times.

**Comment by helicopter88** [comment]
Looks gut to me

File: frontend/flaskserver.py [hunk 1]
```diff
@@ -80,10 +80,22 @@ def get_mime_string(path):
...
 
+@app.after_request
+def add_header(r):
+    """
+    Add headers to both force latest IE rendering engine or Chrome Frame,
+    and also to cache the rendered page for 10 minutes.
+    """
+    r.headers["Cache-Control"] = "no-cache, no-store, must-revalidate"
+    r.headers["Pragma"] = "no-cache"
+    r.headers["Expires"] = "0"
```
Comment (AbraaoMota) [inline frontend/flaskserver.py:95]: I like this bit; however, shouldn't Expires be set to 10 (or 600) for 10 minutes to expire, not 0? Or is 10 mins some standard default?

File: frontend/pem_to_jks.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+openssl pkcs12 -export -out eneCert.pkcs12 -in server.pem
```
Comment (AbraaoMota) [inline frontend/pem_to_jks.sh:1]: Summarised purpose of this file?

File: frontend/webclient/script/professorcontroller.js [hunk 3]
```diff
@@ -287,3 +287,7 @@ theapp.controller('professorController', function($scope) {
   }
 
 });
+
+
+$localstore.ready += 1;
+$localstore.check_ready();
```
Comment (AbraaoMota) [inline frontend/webclient/script/professorcontroller.js:293]: Nice

**Review by AbraaoMota — APPROVED** [review]
Looks gut, just a couple questions

File: frontend/flaskserver.py (hunk 1 above)
Comment (giuliojiang) [reply frontend/flaskserver.py:95]: I don't think it even matters with "no-cache" set.
The reason why I disabled cache is to
* remove some potential problems when debugging stuff and the client doesn't refresh the cache
* firefox doesn't like angularjs applications with cache on, loading is really slow

File: frontend/pem_to_jks.sh (hunk 2 above)
Comment (giuliojiang) [reply frontend/pem_to_jks.sh:1]: take a single server.pem file containing SSL certificate and private key, and generate whatever is needed by the python SSL context (just the certificate and private key separated) and by the java SSL context (which needs things stored in a password-protected keystore in JKS format)

File: frontend/webclient/script/professorcontroller.js (hunk 3 above)
Comment (giuliojiang) [reply frontend/webclient/script/professorcontroller.js:293]: no. It's a hack

**Comment by giuliojiang** [comment]
solved conflicts. Will merge after TeamCity finishes build

**Pull request merged**